spec/controllers/case_management/tax_returns_controller_spec.rb

```diff
@@ -100,54 +104,243 @@
       before { sign_in user }
 
-      it "returns an ok response" do
-        post :edit_status, params: params
-        expect(response).to be_ok
+      it "redirects to the messages tab with a basic flash success message" do
+        post :update_status, params: params
+
+        expect(response).to redirect_to case_management_client_path(id: tax_return.client)
+        expect(flash[:notice]).to eq "Success: Action taken! Updated status."
       end
 
-      it "finds the tax return" do
-        get :edit_status, params: params
+      context "when a new status is submitted" do
+        let(:status) { "intake_needs_assignment" }
 
-        expect(assigns(:tax_return)).to eq(tax_return)
+        it "creates a system note and updates the status" do
+          expect(SystemNote).to receive(:create_status_change_note).with(user, tax_return)
+
+          post :update_status, params: params
+          expect(tax_return.reload.status).to eq("intake_needs_assignment")
+          expect(flash[:notice]).to match "Updated status"
+        end
+      end
+
+      context "when the status is the same as he current status" do
+        let(:status) { "intake_in_progress" }
+
+        it "does not create a system status change note" do
+          expect(SystemNote).not_to receive(:create_status_change_note).with(user, tax_return)
+
+          post :update_status, params: params
+        end
+      end
+
+      context "there is content in the note field" do
+        let(:internal_note) { 'Lorem ipsum note about client tax return' }
+
+        it "saves a note" do
+          expect do
+            post :update_status, params: params
+          end.to change(Note, :count).by(1)
+
+          note = Note.last
+          expect(note.client).to eq tax_return.client
+          expect(note.body).to eq internal_note
+          expect(note.user).to eq user
+
+          expect(flash[:notice]).to match "added internal note"
+        end
+      end
+
+      context "when the note field is blank" do
+        let(:internal_note) { " \n" }
+
+        it "does not save a note" do
+          expect do
+            post :update_status, params: params
+          end.not_to change(Note, :count)
+        end
+      end
+
+      context "when the message body is blank" do
+        let(:message_body) { " \n" }
+        let(:contact_method) { "email" }
+
+        it "does not send a text message nor email" do
+          expect do
+            post :update_status, params: params
+          end.not_to change(OutgoingEmail, :count)
+        end
+      end
+
+      context "when the message body is present" do
+        let(:message_body) { "There's money in the banana stand" }
+
+        context "and the contact method is email" do
+          let(:contact_method) { "email" }
+          let(:mock_mailer) { double }
+          let(:example_now_time) { DateTime.new(2020, 11, 20) }
+
+          before do
+            allow(mock_mailer).to receive(:deliver_later)
```

Comment:
You can use Rails's mailer tests rather than mocking the mailer if you want. See e.g. https://relishapp.com/rspec/rspec-rails/docs/mailer-specs

Not essential, just a FYI.

Reply:
In refactor soon, I wanted to out all the message sending logic in one place so that the controllers wouldn't have to check so many things.

Reply:
I've now included that refactor, but for now I've left the unit tests as they are. If you feel good about it, I'd like to move the unit tests onto the shared controller concern, and then just check that the controllers call the right method to send out emails.

Reply:
+1 to move the tests onto the concern & validate here we do/don't call the method on the concern